src/org/traccar/smpp/SmppClient.java

```diff
@@ -0,0 +1,218 @@
+ * limitations under the License.
+ */
+package org.traccar.smpp;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+import org.traccar.Context;
+import org.traccar.helper.Log;
+
+import com.cloudhopper.commons.charset.CharsetUtil;
+import com.cloudhopper.smpp.SmppBindType;
+import com.cloudhopper.smpp.SmppConstants;
+import com.cloudhopper.smpp.SmppSession;
+import com.cloudhopper.smpp.SmppSessionConfiguration;
+import com.cloudhopper.smpp.impl.DefaultSmppClient;
+import com.cloudhopper.smpp.impl.DefaultSmppSessionHandler;
+import com.cloudhopper.smpp.pdu.SubmitSm;
+import com.cloudhopper.smpp.pdu.SubmitSmResp;
+import com.cloudhopper.smpp.type.Address;
+import com.cloudhopper.smpp.type.RecoverablePduException;
+import com.cloudhopper.smpp.type.SmppChannelException;
+import com.cloudhopper.smpp.type.SmppTimeoutException;
+import com.cloudhopper.smpp.type.UnrecoverablePduException;
+
+public class SmppClient {
+
+    private SmppSessionConfiguration sessionConfig = new SmppSessionConfiguration();
+    private SmppSession smppSession;
+    private DefaultSmppSessionHandler sessionHandler = new ClientSmppSessionHandler(this);
+    private ExecutorService executorService = Executors.newCachedThreadPool();
+    private DefaultSmppClient clientBootstrap = new DefaultSmppClient(executorService, 1);
+
+    private ScheduledExecutorService enquireLinkExecutor;
+    private ScheduledFuture<?> enquireLinkTask;
+    private Integer enquireLinkPeriod;
+    private Integer enquireLinkTimeout;
+
+    private ScheduledExecutorService reconnectionExecutor;
+    private ScheduledFuture<?> reconnectionTask;
+    private Integer reconnectionDelay;
+
+    private String sourceAddress;
+    private int submitTimeout;
+    private String charsetName;
+    private byte dataCoding;
+
+    private byte sourceTon;
+    private byte sourceNpi;
+
+    private byte destTon;
+    private byte destNpi;
+
+    public SmppClient() {
+        sessionConfig.setName("Traccar.smppSession");
+        sessionConfig.setInterfaceVersion(
+                (byte) Context.getConfig().getInteger("sms.smpp.version", SmppConstants.VERSION_3_4));
+        sessionConfig.setType(SmppBindType.TRANSCEIVER);
+        sessionConfig.setHost(Context.getConfig().getString("sms.smpp.host", "localhost"));
+        sessionConfig.setPort(Context.getConfig().getInteger("sms.smpp.port", 2775));
+        sessionConfig.setSystemId(Context.getConfig().getString("sms.smpp.username", "user"));
+        sessionConfig.setPassword(Context.getConfig().getString("sms.smpp.password", "password"));
+        sessionConfig.getLoggingOptions().setLogBytes(false);
+        sessionConfig.getLoggingOptions().setLogPdu(Context.getConfig().getBoolean("sms.smpp.logPdu"));
```

Comment:
Agree, but cloudhopper uses slf4j, traccar uses log4j, there is only log4j -> slf4j but we need opposite.
I'll be really appreciate if you point me a solution.